src/olympia/addons/serializers.py

```diff
@@ -211,6 +212,48 @@ class Meta:
                   'release_notes', 'reviewed', 'url', 'version')
 
 
+class CurrentVersionSerializer(SimpleVersionSerializer):
+    def to_representation(self, obj):
+        # If the add-on is a langpack, and `appversion` is passed, try to
+        # determine the latest public compatible version and replace the obj
+        # with the result. Because of the perf impact, only done for langpacks
+        # in the detail API.
+        request = self.context.get('request')
+        view = self.context.get('view')
+        addon = obj.addon
+        if (request and request.GET.get('appversion') and
+                getattr(view, 'action', None) == 'retrieve' and
+                addon.type == amo.ADDON_LPAPP):
+            obj = self.get_current_compatible_version(addon)
+        return super(CurrentVersionSerializer, self).to_representation(obj)
+
+    def get_current_compatible_version(self, addon):
+        """
+        Return latest public version compatible with the app & appversion
+        passed through the request, or fall back to addon.current_version if
+        none is found.
+
+        Only use on langpacks if the appversion parameter is present.
+        """
+        request = self.context.get('request')
+        try:
+            application = AddonAppQueryParam(request).get_value()
+        except ValueError:
+            raise exceptions.ParseError('Invalid or missing app parameter.')
```

Comment:
I know this is how we've done it in the other places in view.py that uses `AddonAppQueryParam` but, imo, this logic should be defined once in the filter itself, like the other filters do.